sdk/v2/core/logs.go

```diff
@@ -0,0 +1,152 @@
+	// left blank, it is presumed logs are desired for the Worker itself.
+	Job string
+	// Container specifies, by name, a container belonging to the Worker or Job
+	// whose logs are being retrieved. If left blank, a container with the same
+	// name as the Worker or Job is assumed.
+	Container string
+}
+
+// LogStreamOptions represents useful options for streaming logs from a specific
+// container of a Worker or Job.
+type LogStreamOptions struct {
+	// Follow indicates whether the stream should conclude after the last
+	// available line of logs has been sent to the client (false) or remain open
+	// until closed by the client (true), continuing to send new lines as they
+	// become available.
+	Follow bool `json:"follow"`
+}
+
+// LogEntry represents one line of output from an OCI container.
+type LogEntry struct {
+	// Time is the time the line was written.
+	Time *time.Time `json:"time,omitempty"`
+	// Message is a single line of log output from an OCI container.
+	Message string `json:"message,omitempty"`
+}
+
+// LogsClient is the specialized client for managing Logs with the Brigade API.
+type LogsClient interface {
+	// Stream returns a channel over which logs for an Event's Worker, or using
+	// the LogsSelector parameter, a Job spawned by that Worker (or a specific
+	// container thereof), are streamed.
+	Stream(
+		ctx context.Context,
+		eventID string,
+		selector LogsSelector,
+		opts LogStreamOptions,
+	) (<-chan LogEntry, <-chan error, error)
+}
+
+type logsClient struct {
+	*restmachinery.BaseClient
+}
+
+// NewLogsClient returns a specialized client for managing Event Logs.
+func NewLogsClient(
+	apiAddress string,
+	apiToken string,
+	allowInsecure bool,
+) LogsClient {
+	return &logsClient{
+		BaseClient: &restmachinery.BaseClient{
+			APIAddress: apiAddress,
+			APIToken:   apiToken,
+			HTTPClient: &http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						InsecureSkipVerify: allowInsecure,
+					},
+				},
+			},
+		},
+	}
+}
+
+func (l *logsClient) Stream(
+	ctx context.Context,
+	eventID string,
+	selector LogsSelector,
+	opts LogStreamOptions,
+) (<-chan LogEntry, <-chan error, error) {
+	queryParams := map[string]string{}
+	if selector.Job != "" {
+		queryParams["job"] = selector.Job
+	}
+	if selector.Container != "" {
+		queryParams["container"] = selector.Container
+	}
+	if opts.Follow {
+		queryParams["follow"] = "true"
+	}
+
+	resp, err := l.SubmitRequest(
+		ctx,
+		restmachinery.OutboundRequest{
+			Method:      http.MethodGet,
+			Path:        fmt.Sprintf("v2/events/%s/logs", eventID),
+			AuthHeaders: l.BearerTokenAuthHeaders(),
+			QueryParams: queryParams,
+			SuccessCode: http.StatusOK,
+		},
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	logCh := make(chan LogEntry)
+	errCh := make(chan error)
+
+	go l.receiveStream(ctx, resp.Body, logCh, errCh)
+
+	return logCh, errCh, nil
+}
+
+// receiveStream is used to receive log messages as SSEs (server sent events),
```

Comment:
I couldn't find any more references to using SSE - is this exposed to other clients as well? And if so, are SSEs the only streaming option?

Reply:
I should mention SSEs explicitly in the docs for the `steam` function of the `LogsClient` interface.

It is the only option I was thinking of providing currently-- but isn't the only one I built, initially. I had originally _also_ implemented a one-shot `GetLogs(...)`...

You can see with all resource types, I have added a `<Resource>List` object that has fields to support paging. (This, by the way, it modeled after how k8s does it.) The reason for doing that is to avoid OOMs if you try to retrieve a massive number of any sort of resource in one shot. Where logs where concerned, I started to implement paging for `GetLogs(...)` and then started asking myself why I was bothering to do that when log streaming was already a thing.

Let me know if you think there is rationale for still having something like `GetLogs(...)`. (Or let me know if I am misinterpreting your question.)

Reply:
Coincidentally, our log stream works the same way k8's own log streams work.

Reply:
I am mostly asking from the perspective of someone implementing multiple dashboards for getting logs (web and CLI dashboards), where a plain `GET` request is always simpler compared to anything else.

I would explore other server push ways as well (directly using HTTP/2 streams, for example).

Reply:
> directly using HTTP/2 streams, for example

Isn't that what I've done? I modeled the code after the clockstream example on https://http2.golang.org/

Reply:
Err... maybe that isn't super apparent without seeing the server-side implementation.

Here's a link to relevant code in the prototype API server:

https://github.com/krancour/brignext/blob/master/v2/apiserver/internal/core/rest/logs_endpoints.go#L63-L73

Reply:
Ah, that is excellent - this is definitely something we want to revisit with the API server, but this seems reasonable (HTTP/2 would have been my first choice).